### PR TITLE
Expose Device Stats in Clients API

### PR DESCRIPTION
**Pull request by notnoop**
This exposes `DeviceGroupStats` to the client stats API.  A sample client output is found in https://gist.github.com/notnoop/fa95592817bf4bfef935a4b818c8afc2 .

There are some incidental changes in independent commits, best reviewed independently.

### CODE REVIEW

**Comment by notnoop** [comment]
@dadgar this is ready for review again.  Also, I included NodeResources update in here.

File: api/nodes.go [hunk 1]
```diff
@@ -601,6 +604,67 @@ type HostDiskStats struct {
 	InodesUsedPercent float64
 }
 
+// DeviceGroupStats contains statistics for each device of a particular
+// device group, identified by the vendor, type and name of the device.
+type DeviceGroupStats struct {
```
Comment (notnoop) [inline api/nodes.go]: I copied the types here verbatim from plugins package.  Not sure if there are conventions we follow when doing that.

File: client/client.go [hunk 2]
```diff
@@ -1134,6 +1131,23 @@ func (c *Client) updateNodeFromDriver(name string, info *structs.DriverInfo) *st
...
+// updateNodeFromFingerprint updates the node with the result of
+// fingerprinting the node from the diff that was created
+func (c *Client) updateNodeFromDevices(devices []*structs.NodeDeviceResource) {
+	c.configLock.Lock()
+	defer c.configLock.Unlock()
+
+	// Not updating node.Resources: the field is deprecated and includes
+	// dispatched task resources and not appropriate for expressing
+	// node available device resources
+	if !structs.DevicesEquals(c.config.Node.NodeResources.Devices, devices) {
```
Comment (notnoop) [inline client/client.go]: I initially went the Fingerprint route, but then realized that all updates to resources happen independently by type (i.e. cpu vs memory vs disk), so going fingerprint route didn't provide much benefit.

So using `updateNodeFromDevices` directly as it's simpler.  Also confirmed that updates to node are accumulative properly (with a test), and fingerprint manager gets updated node and subsequent fingerprint updates don't unnecessarily update node or stomp values.  So I think this is good.

File: api/nodes.go (hunk 1 above)
Comment (dadgar) [reply api/nodes.go]: Nope this is good

File: vendor/vendor.json [hunk 3]
```diff
@@ -2,439 +2,2705 @@
...
-		{"path":"github.com/prometheus/procfs/xfs","checksumSHA1":"xCiFAAwVTrjsfZT1BIJQ3DgeNCY=","revision":"e645f4e5aaa8506fc71d6edbc5c4ff02c04c46f2","revisionTime":"2017-07-03T10:12:42Z"},
-		{"path":"github.com/rkt/rkt/api/v1","checksumSHA1":"ge4Z0w3QJJYTBqJaK3S+a6MCxzQ=","revision":"3abde24bc284b7ded5784c56b4e8184c28999641","revisionTime":"2017-08-01T12:18:56Z","version":"v1.28.1","versionExact":"v1.28.1"},
-		{"path":"github.com/rkt/rkt/networking/netinfo","checksumSHA1":"4QqLbh9MmajcN6gCx8Er1voiQys=","revision":"5e83d91aafef5f7a38fef62c045e8b57eeeb8bce","revisionTime":"2017-09-20T12:17:54Z"},
-		{"path":"github.com/rs/cors","checksumSHA1":"I778b2sbNN/yjwKSdb3y7hz2yUQ=","revision":"eabcc6af4bbe5ad3a949d36450326a2b0b9894b8","revisionTime":"2017-08-01T07:32:01Z"},
-		{"path":"github.com/ryanuber/columnize","checksumSHA1":"M57Rrfc8Z966p+IBtQ91QOcUtcg=","comment":"v2.0.1-8-g983d3a5","revision":"abc90934186a77966e2beeac62ed966aac0561d5","revisionTime":"2017-07-03T20:58:27Z"},
-		{"path":"github.com/ryanuber/go-glob","checksumSHA1":"6JP37UqrI0H80Gpk0Y2P+KXgn5M=","revision":"256dc444b735e061061cf46c809487313d5b0065","revisionTime":"2017-01-28T0
```
Comment (dadgar) [inline vendor/vendor.json]: run `make vendorfmt`

**Pull request merged**